source/OpenBVE/Parsers/Panel/PanelXmlParser.cs

```diff
@@ -0,0 +1,1629 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml;
+using System.Xml.Linq;
+using OpenBveApi;
+using OpenBveApi.Colors;
+using OpenBveApi.FunctionScripting;
+using OpenBveApi.Interface;
+using OpenBveApi.Math;
+using OpenBveApi.Objects;
+using OpenBveApi.Textures;
+
+namespace OpenBve.Parsers.Panel
+{
+	internal static class PanelXmlParser
+	{
+		// constants
+		private static double StackDistance = 0.000001;
+		/// <remarks>EyeDistance is required to be 1.0 by UpdateCarSectionElement and by UpdateCameraRestriction, thus cannot be easily changed.</remarks>
+		private const double EyeDistance = 1.0;
+
+		/// <summary>Parses a openBVE panel.xml file</summary>
+		/// <param name="PanelFile">The relative path of the panel configuration file from the train</param>
+		/// <param name="TrainPath">The on-disk path to the train</param>
+		/// <param name="Train">The train</param>
+		/// <param name="Car">The car index to add the panel to</param>
+		internal static void ParsePanelXml(string PanelFile, string TrainPath, TrainManager.Train Train, int Car)
+		{
+			// The current XML file to load
+			string FileName = Path.CombineFile(TrainPath, PanelFile);
+			XDocument CurrentXML = XDocument.Load(FileName, LoadOptions.SetLineInfo);
+
+			// Check for null
+			if (CurrentXML.Root == null)
+			{
+				// We couldn't find any valid XML, so return false
+				throw new System.IO.InvalidDataException();
+			}
+
+			IEnumerable<XElement> DocumentElements = CurrentXML.Root.Elements("Panel");
+
+			// Check this file actually contains OpenBVE panel definition elements
+			if (DocumentElements == null)
```

Comment:
I think we probably want to check for count > 0 here?
Otherwise, if valid XML we get a collection with zero nodes.

Reply:
As you say, I would rather check the count> 0 and throw an exception.